### PR TITLE
[AC remat] Skip backward regions that don't need recomputation

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #180535
* __->__ #180534

The remat pass previously errored on any graph with multiple disjoint
backward regions. This is too strict for patterns like chunked loss,
where intermediate chunk_loss.backward() calls create backward regions
that don't depend on any recomputable forward nodes.

Now we collect all backward regions, filter to those that actually need
remat, and only error if multiple regions need it. User-annotated
phase regions (multiple annotations) still always error. The remat
loop itself is unchanged.

Authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@claude Do review.

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 1]
```diff
@@ -57,26 +58,32 @@ def _has_user_phase_annotation(gm: fx.GraphModule) -> bool:
     )
 
 
-def _collect_backward_region_data(
-    gm: fx.GraphModule,
-) -> tuple[bool, int | None, int | None, int]:
+@dataclasses.dataclass
+class BackwardRegions:
```
Comment (soulitzer) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: I feel like this class could just be something like "BackwardRegion" which represents either (bwd_start, len) or (bwd_start, bwd_end)

The other two fields use_phase and nodes are more properties of the entire AC execution / graph.

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 2]
```diff
@@ -132,6 +129,28 @@ def remat_using_tags_for_fwd_loss_bwd_graph(gm: fx.GraphModule) -> fx.GraphModul
 
     force_save_bw_mutation_src(gm)
 
+    def _region_needs_remat(start: int, end: int) -> bool:
+        return any(
+            must_recompute(inp)
+            for node in itertools.islice(gm.graph.nodes, start, end)
+            for inp in node.all_input_nodes
```
Comment (soulitzer) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: Since we were already iterating through nodes when identifying regions, I wonder if this check could just be inline there.

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py [hunk 3]
```diff
@@ -57,26 +58,32 @@ def _has_user_phase_annotation(gm: fx.GraphModule) -> bool:
...
 
 
-def _collect_backward_region_data(
-    gm: fx.GraphModule,
-) -> tuple[bool, int | None, int | None, int]:
+@dataclasses.dataclass
+class BackwardRegions:
+    use_phase: bool
+    nodes: list[fx.Node]
+    regions: list[tuple[int, int]]
```
Comment (aorenste) [inline torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: at a minimum remove `nodes`, add comments
```suggestion
    # True if the user explicitly annotated backward regions via phase="backward".
    use_phase: bool
    # Maximal contiguous runs of backward nodes, 
    # as [start, end) indices into the graph node list.
    regions: list[tuple[int, int]]
```

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py (hunk 2 above)
Comment (aorenste) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: I think must_recompute can only be called after cleanup_recompute_tags...

**Review by aorenste — COMMENTED** [review]
Approving w/ comments

File: torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/_functorch/_activation_checkpointing/remat_using_tags_for_fwd_loss_bwd_graph_pass.py]: We already do it after cleanup_recompute_tags.

**Review by soulitzer — APPROVED** [review]
